compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -109,45 +126,159 @@ impl InternalAudioMixer {
+            .first()
+            .map(|batch| {
+                let duration_secs = start_pts.as_secs_f64() - batch.start_pts.as_secs_f64();
+                let sample_duration_secs = 1.0 / sample_rate as f64;
+                Duration::from_secs_f64(duration_secs.rem_euclid(sample_duration_secs))
+            })
+            .unwrap_or(Duration::ZERO);
+
+        let time_to_sample_count = |duration: Duration| {
+            let sample_count = duration.as_secs_f64() * sample_rate as f64;
+            if (sample_count - sample_count.round()).abs() < max_error.as_secs_f64() {
+                sample_count.round() as usize
+            } else {
+                sample_count.floor() as usize
+            }
+        };
+
+        let last_batch_end_pts = samples.last().map(|sample| sample.end_pts + sample_offset);
+        let samples_iter = samples.into_iter().map(|mut sample| {
+            sample.start_pts += sample_offset;
+            sample.end_pts += sample_offset;
+            sample
+        });
+
+        for (batch_index, input_samples) in samples_iter.enumerate() {
+            let sample_count = samples_in_frame.len();
+            let expected_next_sample_start_pts =
+                start_pts + Duration::from_secs_f64(sample_count as f64 / sample_rate as f64);
+
+            // potentially fill missing spots
+            if expected_next_sample_start_pts + max_error < input_samples.start_pts {
+                let missing_time = input_samples
+                    .start_pts
+                    .saturating_sub(expected_next_sample_start_pts);
+                let missing_samples_count = time_to_sample_count(missing_time);
+                if missing_samples_count < 1 {
+                    warn!(
+                        ?missing_time,
+                        "Distance between samples is higher than expected."
+                    )
+                }
+                samples_in_frame.extend((0..missing_samples_count).map(|_| (0i16, 0i16)))
+            }
+
+            let sample_count = samples_in_frame.len();
+            let expected_next_sample_start_pts =
+                start_pts + Duration::from_secs_f64(sample_count as f64 / sample_rate as f64);
+
+            // check if we need to drop samples at the beginning
+            let mut start_range = 0;
+            if expected_next_sample_start_pts > input_samples.start_pts + max_error {
+                let time_to_remove_from_start =
+                    expected_next_sample_start_pts.saturating_sub(input_samples.start_pts);
+                let samples_to_remove_from_start = time_to_sample_count(time_to_remove_from_start);
+                if batch_index != 0 {
+                    // We should only drop samples in the first batch.
+                    warn!(
+                        "Decoder produced overlapping samples. Dropping {} samples.",
+                        samples_to_remove_from_start
+                    );
+                }
```

Comment:
How can you know that at this point? Those PTS values are changed multiple times across the pipeline. Here, in the mixer, in the queue, maybe in the resampler. How do you know that there is not a bug in any of those places, that caused that? Well, I don't. Logging such warnings at this point can point developers to the wrong place. This is even worse than no message at all.